ocaml/xapi/pool_periodic_update_sync.ml

```diff
@@ -0,0 +1,193 @@
+  let utc_start_of_today =
+    Ptime.of_date_time ((y, m, d), ((0, 0, 0), tz_offset_s)) |> Option.get
+  in
+
+  match frequency with
+  | `daily ->
+      (* schedule in the next day *)
+      let one_day_span = Ptime.Span.of_d_ps (1, 0L) |> Option.get in
+      Ptime.add_span utc_start_of_today one_day_span |> Option.get
+  | `weekly ->
+      let days =
+        let today_wday = Ptime.weekday ~tz_offset_s utc_now |> weekday_to_int in
+        match today_wday < day_configed_int with
+        | true ->
+            day_configed_int - today_wday
+        | false ->
+            day_configed_int + 7 - today_wday
+      in
+      let span = Ptime.Span.of_d_ps (days, 0L) |> Option.get in
+      Ptime.add_span utc_start_of_today span |> Option.get
+
+let next_scheduled_datetime ~delay ~utc_now ~tz_offset_s =
+  let span = Ptime.Span.of_float_s delay |> Option.get in
+  let target = Ptime.add_span utc_now span |> Option.get in
+  let (y, m, d), ((hh, mm, ss), _) = Ptime.to_date_time ~tz_offset_s target in
+  (y, m, d, hh, mm, ss)
+
+let print_next_schedule ~delay ~utc_now ~tz_offset_s =
+  debug "[PeriodicUpdateSync] delay for next update sync: %f seconds" delay ;
+  let y, m, d, hh, mm, ss =
+    next_scheduled_datetime ~delay ~utc_now ~tz_offset_s
+  in
+  debug
+    "[PeriodicUpdateSync] next update sync scheduled at pool time: %d-%d-%d, \
+     %d:%d:%d "
+    y m d hh mm ss
+
+let calc_delay ~utc_now ~utc_start_of_next_sched_day ~seconds_in_a_day =
+  let random_span = Ptime.Span.of_float_s seconds_in_a_day |> Option.get in
+  let utc_next_schedule =
+    Ptime.add_span utc_start_of_next_sched_day random_span |> Option.get
+  in
+  let calc_delay =
+    Ptime.diff utc_next_schedule utc_now |> Ptime.Span.to_float_s
+  in
+  if Xapi_fist.disable_periodic_update_sync_sec_randomness () then
+    calc_delay
+  else
+    (* Make the minimum interval between 2 runs to 2 hours to avoid too close schedules *)
+    Float.max calc_delay update_sync_minimum_interval
+
+let update_sync_delay_for_next_schedule ~__context =
+  let frequency =
+    Db.Pool.get_update_sync_frequency ~__context
+      ~self:(Helpers.get_pool ~__context)
+  in
+  let day_configed =
+    Db.Pool.get_update_sync_day ~__context ~self:(Helpers.get_pool ~__context)
+  in
+  debug
+    "[PeriodicUpdateSync] update_sync_delay_for_next_schedule, frequency=%s, \
+     day_configed=%Ld"
+    (frequency_to_str ~frequency)
+    day_configed ;
+  let day_configed_int = Int64.to_int day_configed in
+  let utc_now = Ptime_clock.now () in
+  let tz_offset_s = Ptime_clock.current_tz_offset_s () |> Option.get in
+  let seconds_in_a_day = seconds_random_within_a_day () in
+  let utc_start_of_next_sched_day =
+    utc_start_of_next_scheduled_day ~utc_now ~tz_offset_s ~frequency
+      ~day_configed_int
+  in
+  let delay =
+    calc_delay ~utc_now ~utc_start_of_next_sched_day ~seconds_in_a_day
+  in
+  print_next_schedule ~delay ~utc_now ~tz_offset_s ;
+  delay
```

Comment:
Updated, but I think it is useful to have some logs here since it will not spam(only once a day for daily schedule, or once a week for weekly schedule), for example, Qin tested update sync these days, and for 2 times, there is no sign (no log from "update sync") that "update sync" had ever be triggered, seems it failed silently. And it had not been reproduced later.